xds/internal/clients/xdsclient/ads_stream.go

```diff
@@ -0,0 +1,803 @@
+	// constructor and are read-only afterwards, and hence can be accessed
+	// without a mutex.
+	transport          clients.Transport       // Transport to use for ADS stream.
+	eventHandler       streamEventHandler      // Callbacks into the xdsChannel.
+	backoff            func(int) time.Duration // Backoff for retries, after stream failures.
+	nodeProto          *v3corepb.Node          // Identifies the gRPC application.
+	watchExpiryTimeout time.Duration           // Resource watch expiry timeout
+	logger             *igrpclog.PrefixLogger
+
+	// The following fields are initialized in the constructor and are not
+	// written to afterwards, and hence can be accessed without a mutex.
+	streamCh     chan clients.Stream // New ADS streams are pushed here.
+	requestCh    *buffer.Unbounded   // Subscriptions and unsubscriptions are pushed here.
+	runnerDoneCh chan struct{}       // Notify completion of runner goroutine.
+	cancel       context.CancelFunc  // To cancel the context passed to the runner goroutine.
+
+	// Guards access to the below fields (and to the contents of the map).
+	mu                sync.Mutex
+	resourceTypeState map[ResourceType]*resourceTypeState // Map of resource types to their state.
+	fc                *adsFlowControl                     // Flow control for ADS stream.
+	firstRequest      bool                                // False after the first request is sent out.
+}
+
+// streamOpts contains the options for creating a new ADS Stream.
+type streamOpts struct {
+	Transport          clients.Transport       // xDS transport to create the stream on.
+	EventHandler       streamEventHandler      // Callbacks for stream events.
+	Backoff            func(int) time.Duration // Backoff for retries, after stream failures.
+	NodeProto          *v3corepb.Node          // Node proto to identify the gRPC application.
+	WatchExpiryTimeout time.Duration           // Resource watch expiry timeout.
+	LogPrefix          string                  // Prefix to be used for log messages.
+}
+
+// newStreamImpl initializes a new StreamImpl instance using the given
+// parameters.  It also launches goroutines responsible for managing reads and
+// writes for messages of the underlying stream.
+func newStreamImpl(opts streamOpts) *streamImpl {
+	s := &streamImpl{
+		transport:          opts.Transport,
+		eventHandler:       opts.EventHandler,
+		backoff:            opts.Backoff,
+		nodeProto:          opts.NodeProto,
+		watchExpiryTimeout: opts.WatchExpiryTimeout,
+
+		streamCh:          make(chan clients.Stream, 1),
+		requestCh:         buffer.NewUnbounded(),
+		runnerDoneCh:      make(chan struct{}),
+		resourceTypeState: make(map[ResourceType]*resourceTypeState),
+	}
+
+	l := grpclog.Component("xds")
+	s.logger = igrpclog.NewPrefixLogger(l, opts.LogPrefix+fmt.Sprintf("[ads-stream %p] ", s))
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+	go s.runner(ctx)
```

Comment:
Same here. We should have a log statement here that the ADS stream is created. This will help even without anything uniquely identifying the stream (because the pointer will uniquely identify it), and this can help track log messages for the duration of the stream.

Reply:
Update. I do see a message when the actual stream is created in `runner`. But I think it wont hurt if we throw a log message here as well. But we should be sure to distinguish it from the message in `runner`.